### ISSUE TITLE
Testing

### ISSUE THREAD

**Issue by LKedward** [ivan-pi/kdtree2#12]
I was looking to add a CI, but I realised the tests might need some work first. There are currently some high-level integration tests, but it would probably be good to add some simpler unit-tests if possible. (This may help diagnose #6, which I also encountered when running the tests.)

It also looks like the tests don't seed the random numbers so they're not repeatable currently.

**Comment by ivan-pi** [comment]
Yes, the tests or rather performance benchmarks from the original author need some work too. 

I can look into the kd-tree test suites of other open-source projects (e.g. [Scipy](https://github.com/scipy/scipy/blob/master/scipy/spatial/tests/test_kdtree.pys)) just to get a feeling what aspects are tested.

The first step will be to test the public routines:

```fortran
  !-------------DATA TYPE, CREATION, DELETION---------------------
  public :: kdkind
  public :: kdtree2, kdtree2_result, tree_node, kdtree2_create, kdtree2_destroy
  !---------------------------------------------------------------
  !-------------------SEARCH ROUTINES-----------------------------
  public :: kdtree2_n_nearest, kdtree2_n_nearest_around_point
  ! Return fixed number of nearest neighbors around arbitrary vector,
  ! or extant point in dataset, with decorrelation window.
  !
  public :: kdtree2_r_nearest, kdtree2_r_nearest_around_point
  ! Return points within a fixed ball of arb vector/extant point
  !
  public :: kdtree2_sort_results
  ! Sort, in order of increasing distance, rseults from above.
  !
  public :: kdtree2_r_count, kdtree2_r_count_around_point
  ! Count points within a fixed ball of arb vector/extant point
  !
  public :: kdtree2_n_nearest_brute_force, kdtree2_r_nearest_brute_force
  ! brute force of kdtree2_[n|r]_nearest
  !----------------------------------------------------------------
```

I can put together some tests to check the distances and indexes of the points for consistency. In case of the radius search they can test the points found are indeed within the radius (or compare against the brute force search).

Testing the result sorting is also straightforward.

At present the kd-tree interfaces assumes the tree is a pointer variable. I would like to remove this in the future to simplify usage. I am not too motivated to write tests for the data type creation and deletion routines, knowing the API might change soon. On the other hand having tests and refactoring them along with the core routines, could help inform the redesign process.
